internal/controllers/clusterclass/clusterclass_controller.go

```diff
@@ -299,18 +307,32 @@ func (r *Reconciler) reconcileVariables(ctx context.Context, s *scope) (ctrl.Res
 			if patch.External == nil || patch.External.DiscoverVariablesExtension == nil {
 				continue
 			}
+
 			req := &runtimehooksv1.DiscoverVariablesRequest{}
 			req.Settings = patch.External.Settings
 
-			resp := &runtimehooksv1.DiscoverVariablesResponse{}
-			err := r.RuntimeClient.CallExtension(ctx, runtimehooksv1.DiscoverVariables, clusterClass, *patch.External.DiscoverVariablesExtension, req, resp)
-			if err != nil {
-				errs = append(errs, errors.Wrapf(err, "failed to call DiscoverVariables for patch %s", patch.Name))
-				continue
+			// Considering DiscoverVariables is expected to return a "static" response + usually there are few extensions in a cluster,
+			// we temporarily cache the response to improve performances in case there are many CC using the same RT/the same settings.
+			// This also mitigates spikes when CC re-sync happens or when changes to the ExtensionConfig are applied.
+			e := callCacheEntry{
+				Extension: *patch.External.DiscoverVariablesExtension,
+				Settings:  patch.External.Settings,
 			}
-			if resp.Status != runtimehooksv1.ResponseStatusSuccess {
-				errs = append(errs, errors.Errorf("patch %s returned status %q with message %q", patch.Name, resp.Status, resp.Message))
-				continue
+			var ok bool
+			var resp *runtimehooksv1.DiscoverVariablesResponse
+			if resp, ok = r.callCache.GetByKey(e.Key()); !ok {
```

Comment:
This would allow to bypass the namespace selector of an extensionconfig. Should we add the cache on a deeper layer (somewhere around https://github.com/kubernetes-sigs/cluster-api/blob/401af1c7ddc6db552600bb1cae9a6195d8623fba/internal/runtime/client/client.go#L291)?

Also: maybe add the generation of the extensionconfig to the cache entry key?

Reply:
> This would allow to bypass the namespace selector of an extensionconfig. Should we add the cache on a deeper layer (somewhere around

Nice catch! Done

> Also: maybe add the generation of the extensionconfig to the cache entry key?

I agree. Maybe even the resourceVersion?